utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java

```diff
@@ -0,0 +1,378 @@
+            httpReq.addHeader(ACCEPT, APPLICATION_JSON);
+            httpReq.addHeader(HTTP.CONTENT_TYPE, APPLICATION_JSON);
+            String encoding = basicAuth(username, password);
+            httpReq.addHeader("Authorization", encoding);
+            httpReq.setEntity(new StringEntity(jsonToSend.toString()));
+        } catch (URISyntaxException | UnsupportedEncodingException e) {
+            throw new RedfishException(String.format("Failed to create URI for POST request [URL: %s] due to exception.", url), e);
+        }
+
+        HttpClient client = null;
+        if (ignoreSsl) {
+            try {
+                client = ignoreSSLCertValidator();
+            } catch (NoSuchAlgorithmException | KeyManagementException e) {
+                throw new RedfishException(String.format("Failed to handle SSL Cert validator on POST request [URL: %s] due to exception.", url), e);
+            }
+        } else {
+            client = HttpClientBuilder.create().build();
+        }
+
+        try {
+            return client.execute(httpReq);
+        } catch (IOException e) {
+            throw new RedfishException(String.format("Failed to execute POST request [URL: %s] due to exception.", url, e));
+        }
+    }
+
+    /**
+     *  Returns the proper URL path for the given Redfish command ({@link RedfishCmdType}).
+     */
+    private String getRequestPathForCommand(RedfishCmdType cmd, String resourceId) {
+        switch (cmd) {
+        case GetSystemId:
+            return SYSTEMS_URL_PATH;
+        case GetPowerState:
+            if (StringUtils.isBlank(resourceId)) {
+                throw new RedfishException(String.format("Command '%s' requires a valid resource ID '%s'.", cmd, resourceId));
+            }
+            return String.format("%s%s", SYSTEMS_URL_PATH, resourceId);
+        case ComputerSystemReset:
+            if (StringUtils.isBlank(resourceId)) {
+                throw new RedfishException(String.format("Command '%s' requires a valid resource ID '%s'.", cmd, resourceId));
+            }
+            return String.format("%s%s%s", SYSTEMS_URL_PATH, resourceId, COMPUTER_SYSTEM_RESET_URL_PATH);
+        default:
+            throw new RedfishException(String.format("Redfish client does not support command '%s'.", cmd));
+        }
+    }
+
+    /**
+     * Validates the host address. It needs to be either a valid host domain name, or a valid IP address (IPv6 or IPv4).
+     */
+    protected String validateAddressAndPrepareForUrl(String hostAddress) {
+        if (NetUtils.isValidIp6(hostAddress)) {
+            return String.format("[%s]", hostAddress);
+        } else if (NetUtils.isValidIp4(hostAddress)) {
+            return hostAddress;
+        } else if (InternetDomainName.isValid(hostAddress)) {
+            return hostAddress;
+        } else {
+            throw new RedfishException(String.format("Redfish host address '%s' is not a valid IPv4 or IPv6 address", hostAddress));
```

Comment:
This Exception doesn't say what the code does. A valid hostname is also accepted. Not only IPv4 or IPv6 addresses.

Reply:
Good point, @wido! I will update this soon

Reply:
@wido code has been updated to `Redfish host address '%s' is not a valid IPv4/IPv6 address nor a valid domain name.`